lib/services/event_service.dart

```diff
@@ -2,21 +2,84 @@ import 'package:cloud_firestore/cloud_firestore.dart';
 import 'package:project_fomo/models/event.dart';
 
 class EventService {
-  final CollectionReference _eventCollection =
-      Firestore.instance.collection('events');
-
   final String _userId;
+  DocumentSnapshot lastDocument;
 
   EventService(this._userId);
 
-  Future<List<Event>> getAllEvents() async {
-    QuerySnapshot eventSnapshot = await _eventCollection.getDocuments();
+  Future<Map<String, List<Event>>> getEventsByCategory() async {
```

Comment:
Is it not possible to get categorized events with a single firestore query?  Wasn't that the reason cloud functions were used in the first place?  A query count that grows with the number of categories is a pretty big bottleneck.

Reply:
Yes, cloud functions were the first proposed solution to your concern by attaching tags to all events. You could pull everything down in one query, sort/filter, and then populate Discover Page categories. I switched to this new method for the following reasons:

1)  **With multiple queries, categories don't have to wait on each other to populate.
2) Cloud functions are very much overkill for the amount of data/categories we are working with. Not enough data or categories to justify an optimization like this.
3) The complexity of the sorting/filtering

Reply:
^ extension to point 3: Occams' Razor. I like the multi query design more because of its simplicity.

Reply:
Filtering would be pretty easy.  Once you query all the events with one call you could do the following:  Popular you'd pass a filter that checks if the goingCount/interestedCount is greater than a certain number.  Upcoming you'd pass a filter that compares current time to event time and gets rid of events out of a certain range.  Featured we could manually add in the code, getting rid of all load time for featured.  Cloud functions were supposed to make it so not only did we not need multiple queries, but also no filtering on the client side.  This way is much slower than both the other methods, and if we add more categories which we will, it gets slower and slower.

Reply:
I agree that it wouldn't scale for the reasons mentioned above but our only categories are Featured, Popular, Upcoming, and soon Friends. That is four calls total with no intention of adding others. No need to build something to scale if it won't, therefore I think simplicity wins here.

As for complexity of the filtering, the initial filtering of the first request isn't too bad as you said. It gets a little more twisted when you consider ViewAll. The Discover Page would be querying all categories at once, while the ViewAll page has to have a seperate query per category because it is only trying to display the "view all" of a single category. 

Overall, I see what you are saying but it would be like using a power drill to unscrew a PS4 controller.

Reply:
I don't think adding filtering as a speed improvement update is off the table for the future, just not the most immediate thing that needs attention. I think we can merge this working implementation and make a note that a future speed improvement could be made here.

Reply:
Yeah we can merge and make a note of it.